Implement stack walking (backtrace)

rasky opened the pull request:
Quoting the docs:

```
This module implements two POSIX/GNU standard functions to help walking
the stack and providing the current execution context: backtrace() and
backtrace_symbols().
 
The functions have an API fully compatible with the standard ones. The
implementation is however optimized for the MIPS/N64 case, and with
standard compilation settings. See the documentation in backtrace.c
for implementation details.
 
You can call the functions to inspect the current call stack. For
a higher level function that just prints the current call stack
on the debug channels, see #debug_backtrace.
```

An optimized symbol table is generated by a new tool (n64sym), which is then hooked up to the build system (n64.mk). The symbol table is then embedded in the ROM as rompak.

n64sym analyzes standard DWARF debugging symbol in the ELF file using tools from the toolchain, so that it doesn't have to embed a DWARF parser and is thus hopefully very solid to feature evolutions of these (ever changing) standards. It is quite fast (~100ms for a standard C ROM, ~1 sec for huge C++ ROMs that take forever to link). The symbol table is reasonably compact (100-200K up to 1M for huge ROMs) and is never loaded into ROM, but fully queried directly from ROM. Obviously it can be removed from production builds (backtracing still works without symbol table, but of course only hexadecimal addresses will be printed).

This feature has been extensively tested in the unstable channel and is quite stable and incredibly useful during development. This PR brings in the core functionality, it doesn't hook it up yet to assertions or exceptions.

rasky committed:
tools: start creating common library of functions to help tool programming

For now, add a simple polyfill.h to fill some mingw32 gap wrt modern
glibc/posix, and utils.h which includes the same utils.h used internally
by libdragon with a few common macros.

rasky committed:
tools: add new n64sym tool

This tool will be the basis of the upcoming backtracing support. It
parses debug symbols from an ELF file (using the toolchain's GNU tools
such as objdump) and produces a compact symbol file (extension: .sym)
which is meant to be queried at runtime to symbolize a backtrace.

To make the file as small as possible, the sym file only contains
symbols for function call points. Moreover, data is efficiently stored
so that it can be queried directly from ROM.

The tool runs very quickly on modern computers. Normally it doesn't take
more than 100/200 ms, and even for huge C++ projects it only takes about
one second.

The tool relies on the stb_ds single file library for data structures
(specifically, hash tables and resizable arrays), and
sheredom/subprocess.h, which is a library to portably write code that
spawns an external process (unfortunately, popen() is several limited
on some implementations).

rasky committed:
n64.mk: build with debug symbols and bundle a .sym file in the ROM

This commit runs n64sym by default on all binaries (after building
them with debug symbols), and packs the produced symbol file in the
rompak.

rasky committed:
backtrace: add API to walk the stack, symbolizing functions.

This commit adds the core backtrace functionality for implementing
stack walking.

We implement the standard POSIX API (backtrace() and backtrace_symbols())
plus a more efficient backtrace_symbols_cb() that uses a callback to
avoid memory allocations.

Since it would be too cumbersome to rely on modern DWARF/STABS metadata,
this implementation relies on binary code analysis to infer the shape
of the stack frame and whether frame pointers are used. Symbolization
is then made via the symbol table generated by n64sym and embedded
in the binary.

We also implemented decent coverage by tests on the backtracing
functionality, though new corner cases are always being discovered.

At the topmost API level, we added a debug_backtrace() function that can
be freely called by users to get a backtrace at any point in the code.

This commit does not hook up backtracing to exceptions or assertions yet.

rasky committed:
backtrace: test misaglined addresses and TLB misses separately

anacierdem commented on n64.mk:
```diff
@@ -32,14 +32,15 @@ N64_CHKSUM = $(N64_BINDIR)/chksum64
 
 N64_CFLAGS =  -march=vr4300 -mtune=vr4300 -I$(N64_INCLUDEDIR)
-N64_CFLAGS += -falign-functions=32 -ffunction-sections -fdata-sections
+N64_CFLAGS += -falign-functions=32 -ffunction-sections -fdata-sections -g
```
Is it ok to have this enabled by default? Would it be overkill to only emit debug information and symbol file only if `NDEBUG` is not set? I don't see any reason not to but not sure if someone may object. Though, we can always change this later so not important ATM...

rasky replied:
It can be done, but we can't tie it to NDEBUG (which is a C preprocessor symbol but not a Makefile symbol). We should introduce some Makefile-level variable to control this.

rasky replied:
The important part I think is that the code is always considering symbols an optional part, and is ready to work without it. Obviously, symbolization can't work in that case, but you would still be able to get hexadecimal backtraces.

anacierdem replied:
Yeah we may introduce it in the future this is ok for now.

anacierdem commented on src/backtrace_internal.h:
```diff
@@ -0,0 +1,43 @@
+    BT_FUNCTION,                ///< Regular function with a stack frame
+    BT_FUNCTION_FRAMEPOINTER,   ///< The function uses the register fp as frame pointer (normally, this happens only when the function uses alloca)
+    BT_EXCEPTION,               ///< This is an exception handler (inthandler.S)
+    BT_LEAF                     ///< Leaf function (no calls), no stack frame allocated, sp/ra not modified
```
I suppose the tail calls will get into this category as well?

rasky replied:
Tail-calling is a property of how you call a function, it's not a property of the function itself. Any function of the above types might tailcall another function of any of the above types.

anacierdem replied:
Actually what I meant was a recursive tailcall. In that case it will just call itself without pushing anything to the stack. So shouldn't be any different than a leaf node IINM?

rasky replied:
Yes. In general a recursive tail call is handled by the compiler transforming the code in a loop, so at the assembly level you can't really tell it was written as recursive tail call. That's because a real tail call means popping the whole stack frame and then jumping, but in the recursive case it would be a waste of time to pop and push the same stack frame.

anacierdem commented on src/backtrace_internal.h:
```diff
@@ -0,0 +1,43 @@
+typedef struct {
+    bt_func_type type;       ///< Type of the function
+    int stack_size;          ///< Size of the stack frame
+    int ra_offset;           ///< Offset of the return address in the stack frame
```
I think it is the offset from the top of the frame. Might be nice to include that information as well.

rasky replied:
Done.

anacierdem commented on src/backtrace_internal.h:
```diff
@@ -0,0 +1,43 @@
+ * @param size          Size of the buffer
+ * @return char*        Pointer to the return string. This is within the provided
+ *                      buffer, but not necessarily at the beginning because of DMA
+ *                      alignment constraints.
```
So the size must be at least some amount larger than the potential output, is there a rule of thumb here?

rasky replied:
__symbolize will truncate the symbol to fit the buffer anyway, so any number would do. Consider that n64sym has a maximum size of symbols that can be specified from the command line (default 64), and I think you want to have space for a \0 plus some alignment at the beginning (I think 1 byte should suffice).

I think that if you have 64-ish characters available, you can simply pass a buffer of 128 chars to be on the safe side anyway. These numbers are small enough that it shouldn't be a problem.

anacierdem commented on src/debug.c:
```diff
@@ -70,6 +70,8 @@ static char sdfs_logic_drive[3] = { 0 };
 static void (*debug_writer[3])(const uint8_t *buf, int size) = { 0 };
 
+/** @brief internal backtrace printing function */
+void __debug_backtrace(FILE *out, bool skip_exception);
```
Shouldn't this be static?

rasky replied:
True, but it's going to be used by the new exception handler, in a followup.

anacierdem commented on tools/common/polyfill.h:
```diff
@@ -0,0 +1,74 @@
+#include <stdlib.h>
+#include <string.h>
+
+size_t getline(char **lineptr, size_t *n, FILE *stream) {
```
Can we have a reference if we are using these from somewhere else?

rasky replied:
Done. I've also took the chance to get the improved version I found as this one was buggy.

anacierdem commented on tools/n64sym.c:
```diff
@@ -0,0 +1,487 @@
+    }
+
+    fwrite("SYMT", 4, 1, out);
+    w32(out, 2); // Version
```
What happened to version 1 :)

rasky replied:
The data layout was different and only allowed for 16bit offsets into string table, which is not enough for large C++ codebases. It's in the history of the unstable branch, but we're directly bringing version 2 to trunk.

anacierdem replied:
As the one on the trunk is the actual release I think we can simply reset but that's ok if you want them to be compatible

anacierdem commented on tools/n64sym.c:
```diff
@@ -0,0 +1,487 @@
+        return 1;
+    }
+
+    if (!n64_inst) {
```
Is the check necessary? Seems like it is only set once?

rasky replied:
No you're right, it's a leftover. Removed

anacierdem commented on tools/n64sym.c:
```diff
@@ -0,0 +1,487 @@
+            symbol_add(elf, addr, false);
+        }
+    }
+    free(line);
```
If we are free ing line, it might be also nice to do so for cmd?

rasky replied:
Done.

anacierdem commented on tools/n64sym.c:
```diff
@@ -0,0 +1,487 @@
+        // Remove the trailing backslash if any. On some system, running
+        // popen with a path containing double backslashes will fail, so
+        // we normalize it here.
+        n64_inst = strdup(n64_inst);
```
Might be nice to free things to be a good citizen.

rasky replied:
in this specific case, I resist the urge as this is a single allocation in main, and the free would be done at the end of main so it's not really important, as the process is exiting anyway.

anacierdem replied:
I agree the only thing concerning me was the probability of someone copying it in the future but doesn't seem likely either. Maybe the best opton is just commenting this intent?

rasky replied:
I extracted the code into a `n64_toolchain_dir()` function in common. This allows to easily reuse it in other tools and make sure we don't duplicate the memory leak.

anacierdem commented on tools/n64sym.c:
```diff
@@ -0,0 +1,487 @@
+
+    walign(out, 16);
+    w32_at(out, symtable_off, ftell(out));
+    for (int i=0; i < stbds_arrlen(symtable); i++) {
```
Can we have a reference for the layout of the file? I think at least pointing to the structs at backtrace.c would be nice

rasky replied:
Add references in comments.

anacierdem commented on tools/n64sym.c:
```diff
@@ -0,0 +1,487 @@
+    int addrtable_off = w32_placeholder(out);
+    w32(out, stbds_arrlen(symtable));
+    int symtable_off = w32_placeholder(out);
+    w32(out, stbds_arrlen(symtable));
```
This is symtable length written the second time?
Ok, I later found out why in `backtrace.c` but still I'm not sure why

rasky replied:
So basically the address table contains the address for each symbol in the symbol table.

Looking at the code right now, there's no big win anymore in keeping the address separated from the symbol itself. It used to be more useful for alignment / lookups / dmas reasons but I don't see such a strong benefit anymore. We might evaluate merging the address table and the symbol table to simplify a bit readability of the code.

anacierdem replied:
Is it because the symbol entries are constant length now?

rasky replied:
I don't remember exactly, the code underwent two or three refactorings. Anyway, it could be further improved now but I don't have a rush to do it, as it works quite well.

anacierdem commented on src/backtrace.c:
```diff
@@ -0,0 +1,682 @@
+ * 
+ * To perform backtracing, we scan the code backward starting from the return address
+ * of each frame. While scanning, we note some special instructions that we look
+ * for. The two main instructions that we look for are `sd ra, offset(sp)` which is
```
`sd ra, offset(sp)` does not exist in a leaf node right? So do we simply omit leaf stacks?
I think the later part:
> so that it's always possible to infer the function a certain address belongs to

also explains why we don't need them.

rasky replied:
well in general a leaf function can never be part of a stack trace by definition (as they don't call anything), with the only exception if an exception/interrupt happens in the middle of it. In that case, we can find a leaf function in the stack trace which is indeed a bit of a problem as it's not easy to understand where the function begins. So we handle that with some heuristics and through the symbol map (if present). See the code that handles BT_LEAF.

anacierdem commented on tests/backtrace.S:
```diff
@@ -0,0 +1,147 @@
+
+    # This file contains functions used by test_backtrace.c to test
+    # the backtrace analysis code. The code of this functions is not
+    # run, but just scanned to extract the stack frame information.
```
Is this generated? If so, it would be great to include instructions/source code

rasky replied:
No it's just random copy and paste from disassembly of libdragon code. The code isn't even supposed to make sense, I have just copied and pasted together different snippets of assembly to recreate "normal looking" code for the several testcases.

anacierdem commented on tools/n64sym.c:
```diff
@@ -0,0 +1,487 @@
+    pclose(disasm);
+}
+
+void compact_filenames(void)
```
This is currently unused

rasky replied:
Correct. I wanted to keep it around as I wanted to improve it sooner or later (as explained in the TODO), but I can remove it from trunk if you want.

anacierdem replied:
If we are going to use it soon enough, we can leave. Let's remove o/w

rasky replied:
I have removed this and cherry-picked another follow up change from unstable to remove absolute paths via `-ffile-prefix-map`.

anacierdem commented on src/backtrace.c:
```diff
@@ -0,0 +1,682 @@
+ */
+#define FUNCTION_ALIGNMENT      32
+
+#define MAX_FILE_LEN 120        ///< Maximum length of a file name in a backtrace entry
```
There is no limit on n64sym, why we have limits here?

rasky replied:
Right, this is not needed anymore, we can use VLAs to workaround this at least for the backtrace_symbols_cb case, so that it can handle any size.

For `backtrace_symbols()` (which is the standard API we implement for completeness but it's not the best one to use), since we want to avoid too many allocations, it's better to allocate once with a large enough size, so I moved the constants there.

anacierdem commented on src/backtrace.c:
```diff
@@ -0,0 +1,682 @@
+ *   the source file name and line number, and the binary offset of the symbol within the containing
+ *   function.
+ * * String table: This tables can be thought as a large buffer holding all the strings needed by all
+ *   symbol entries (function names and file names). Each symbol entry stores a string as an index
```
>Each symbol entry stores a string...

is this a typo? I thing should be "an offset" instead of "a string" IINM

rasky replied:
it was:
```Each symbol entry stores a string as an index within the symbol table```
I changed to:
```Each symbol entry stores a string as an offset within the symbol table```

Is this OK?

anacierdem replied:
Ah, yes I think this resolves the confusion for me

anacierdem commented on src/backtrace.c:
```diff
@@ -0,0 +1,682 @@
+ * * String table: This tables can be thought as a large buffer holding all the strings needed by all
+ *   symbol entries (function names and file names). Each symbol entry stores a string as an index
+ *   within the symbol table and a length. This allows to reuse the same string (or prefix thereof)
+ *   multiple times. Notice that strings are not null terminated in the string table.
```
Do we also optimize for substrings?

rasky replied:
The file format would allow that, true. 

The limit is on n64sym, because it becomes quadratic to search for a matching substring in an existing table. So in n64ysm we create a hash map with all prefixes (not all possible substrings), and then look that up. 

I think we could improve it adding to the hashmap some special substrings that we feel are more important than others. Assuming a C symbol like `symbol_add`, we could add to the hash map the word `add` (so any word separated by `_`, or even camel case) as it's more likely for a symbol `add` to exist compared to a symbol called `bol_a` (a random substring).

Anyway, I measured that, even using a full substring search, the gain in size is really minimal, as it rarely happens, so I didn't spend too much time on it.

anacierdem replied:
Yeah, I don't expect much difference. Just was curious b/c gcc seem to do it for sreing constants. In that case they all anchor from the trailing `\0` so probably it is a much smaller search space

anacierdem commented on src/backtrace.c:
```diff
@@ -0,0 +1,682 @@
+    uint32_t addrtab_off;   ///< Offset of the address table in the file
+    uint32_t addrtab_size;  ///< Size of the address table in the file (number of entries)
+    uint32_t symtab_off;    ///< Offset of the symbol table in the file
+    uint32_t symtab_size;   ///< Size of the symbol table in the file (number of entries); always equal to addrtab_size.
```
Is this for "just in case"?

rasky replied:
Yeah, I figured it out that keeping the header more generic wasn't going to be a problem.

anacierdem replied:
On second thought what about "reserving" a few bytes in the header as well? OTOH it is a matter of a rebuild but that extra length made me think about this as well

rasky replied:
It's actually not necessary because anything after the header is accessed via offsets specified in the header itself. That's actually by design.

anacierdem replied:
I mean in case we need to store anything else in the header in the future. I didn't get your point?

anacierdem replied:
Hmm I think I get it now, we will simply shift everything if we need any space.

anacierdem commented on src/backtrace.c:
```diff
@@ -0,0 +1,682 @@
+{
+    if (addr >= (uint32_t)inthandler && addr < (uint32_t)inthandler_end) {
+        // Special case exception handlers. This is just to show something slightly
+        // more readable instead of "notcart+0x0" or similar assembly symbols
```
hmm, maybe we can include the subsymbols as well?

rasky replied:
Honestly they are just local symbols, they're not really expected to be exported. It's more of a side effect of how gas works. If you check unstable, we already changed entrypoint.S to use the syntax `.Lskip_dma` (which is how local labels should be called), so they wouldn't appear anymore in the stacktraces.

I think even if we remove all labels from stacktraces, we would end up with a single symbol called `inthandler`; as much as we could rename it, I think it does provide more clear context to the user using a special symbol like `<EXCEPTION HANDLER>`.

anacierdem reviewed (commented):
This looks awesome! And I know it works on really weird C++ codebases :) Thank you!
Dropped a few comments. Didn't finish `backtrace.c` yet, but wanted to send this so that you can go over them in the meantime.

rasky committed:
tools: improved getline and add a reference from the source

rasky committed:
backtrace: improve comments

rasky committed:
backtrace: remove hardcoded buffer limits in backtrace_symbols_cb

rasky committed:
n64sym: cleanups

anacierdem commented on src/backtrace.c:
```diff
@@ -0,0 +1,680 @@
+        int idx = 0;
+        addrtable_entry_t a = symt_addrtab_search(&symt, addr, &idx);
+        while (!ADDRENTRY_IS_FUNC(a))
+            a = symt_addrtab_entry(&symt, --idx);
```
Maybe we can assert if idx goes down to 0, just in case.

rasky replied:
Done

anacierdem commented on src/backtrace.c:
```diff
@@ -0,0 +1,680 @@
+ * @param func_start                  Start of the function being analyzed. This is optional: the heuristic can work
+ *                                    without this hint, but it is useful in certain situations (eg: to better
+ *                                    walk up after an exception).
+ * @param exception_ra                If != NULL, this function was interrupted by an exception. This variable
```
I am not sure how the exception ra relates to parsing a random function? Shouldn't we be able to analyze the function irrespective of the current execution context?
Also see my later comment, instead of introducing this like a flag, we can just store the trace and merge them later.

anacierdem commented on src/backtrace.c:
```diff
@@ -0,0 +1,680 @@
+
+    // Start from the backtrace function itself. Put the start pointer somewhere after the initial
+    // prolog (eg: 64 instructions after start), so that we parse the prolog itself to find sp/fp/ra offsets.
+    ra = (uint32_t*)backtrace_foreach + 64;
```
Is there a way to use the actual length of this function somehow 🤔

rasky replied:
I guess we could find a way, but at the end of the day we only need to walk through the prologue of the function, we don't need to go to the end. As long as the function is at least 64 opcode, this should work. Otherwise, it becomes more complicated and requires hooking up with the build system.

anacierdem commented on src/backtrace.c:
```diff
@@ -0,0 +1,680 @@
+                if (!func.fp_offset) {
+                    debugf("backtrace: framepointer used but not saved onto stack at %p\n", ra);
+                } else {
+                    // Use the frame pointer to refer to the current frame.
```
...to the current stack?

rasky replied:
Normally, the stack pointer is used to access the frame. When there is a framepointer, though, we must use that one instead as the stack pointer has probably been moved further down. So I think the sentence is correct.

anacierdem commented on src/backtrace.c:
```diff
@@ -0,0 +1,680 @@
+                func_start = 0;
+                break;
+            case BT_EXCEPTION: {
+                // Exception frame. We must return back to EPC, but let's keep the
```
Would it be easier if we trace back from the last ra instead of EPC as if the exception has never occurred and then "append" the exception trace? Then we wouldn't need `exception_ra`. That would look something like:
- create stack trace A
- create stack trace B
- oops, this was an exception, temporarily store [A, B] and start walking back from EPC
- create stack trace C
- create stack trace D
- No more traces, append A, B to C, D

Current `exception_ra` confused me a lot. OTOH, we know this is working so that's ok if you don't think this would be useful.
Also please let me know if I am way of in my theory, you are the one who knows the code inside out.

anacierdem commented on src/backtrace.c:
```diff
@@ -0,0 +1,680 @@
+                }
+
+                // The next frame might be a leaf function, for which we will not be able
+                // to find a stack frame. It is useful to try finding the function start.
```
I didn't exactly get why this is useful only when we are in an exception frame? Considering it is only being used in tandem with `exception_ra`, maybe we won't need it? This part really confused me. I really feel like the logic around the two should be simpler, but I don't know how exactly they relate.

anacierdem commented on src/backtrace.c:
```diff
@@ -0,0 +1,680 @@
+                }
+            }   break;
+            case BT_LEAF:
+                ra = exception_ra - 2;
```
This won't be necessary if we get rid of `exception_ra` as well

anacierdem reviewed (commented):
I have concerns around whether we need `exception_ra` or not. All looks good otherwise. Thank you again for the immense work you put into this, just to make people happy :)

rasky committed:
backtrace: improve __bt_analyze_func clarify and documentation

anacierdem commented on src/backtrace.c:
```diff
@@ -330,7 +330,38 @@ char* __symbolize(void *vaddr, char *buf, int size)
+ *  * The 0.01% of the functions that do not have a stack frame but appear in the call stack are leaf
+ *    functions interrupted by exceptions. Leaf functions pose two important problems: first, $ra is
+ *    not saved into the stack so there is no way to know where to go back. Second, there is no clear
+ *    indication where the function begins (as we normally stops analysis when the see the stack frame
```
typo: `when the see`

rasky replied:
Done

anacierdem commented on src/backtrace.c:
```diff
@@ -0,0 +1,720 @@
+ *                                    This is normally the point where a JAL opcode is found, as we are walking
+ *                                    up the call stack.
+ * @param func_start                  Start of the function being analyzed. This is optional: the heuristic can work
+ *                                    without this hint, but it is useful in certain situations (eg: to better
```
It is actually not used anywhere else for now, but considering it may be used in the future and it is not the same thing as `from_exception` I think it is ok to have it as a separate parameter. Just thinking out loud.

rasky replied:
Yeah I thought the same, for now I'd just leave it separate

anacierdem reviewed (commented):
Looks much better now. I have resolved all points that I understand after the changes :)

rasky committed:
n64.mk: avoid leaking absolute paths by default

This also affects symbols in backtraces, reducing a bit the symbol file
and showing more compact paths on screen.

rasky committed:
n64sym: extract toolchain discovery function

rasky committed:
n64sym: remove dead code

rasky committed:
backtrace: add assert to sanitize uses of symt_addrtab_entry
